### ISSUE TITLE
install/update using --explain 

### ISSUE THREAD

**Issue by MSP-Greg** [rubygems/rubygems#1992]
I'm having a problem.

`--explain` option for `gem install` and `gem update` gives odd results.  I recall this working in a manner that seemed correct in the past.

1. With a specified gem name, it always installs/updates 
2. Without a name, it shows `Nothing to update`, even though a few gems have updates available.

This issue is related to:

  - [ ] Network problems
  - [ ] Installing a library
  - [ ] Publishing a library
  - [x] The command line `gem`
  - [ ] Other

Here are my current environment details:

```
RubyGems Environment:
  - RUBYGEMS VERSION: 2.6.12
  - RUBY VERSION: 2.5.0 (2017-08-25 patchlevel -1) [x64-mingw32]
  - INSTALLATION DIRECTORY: E:/r_builds/25_2017-08-25_59654/ruby25_64/lib/ruby/gems/2.5.0
  - USER INSTALLATION DIRECTORY: C:/Users/Greg/.gem/ruby/2.5.0
  - RUBY EXECUTABLE: E:/r_builds/25_2017-08-25_59654/ruby25_64/bin/ruby.exe
  - EXECUTABLE DIRECTORY: E:/r_builds/25_2017-08-25_59654/ruby25_64/bin
  - SPEC CACHE DIRECTORY: C:/Users/Greg/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: C:/ProgramData
  - RUBYGEMS PLATFORMS:
    - ruby
    - x64-mingw32
  - GEM PATHS:
     - E:/r_builds/25_2017-08-25_59654/ruby25_64/lib/ruby/gems/2.5.0
     - C:/Users/Greg/.gem/ruby/2.5.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
     - "install" => "--no-document --env-shebang --platform ruby -- --use-system-libraries"
     - "update" => "--no-document --env-shebang --platform ruby -- --use-system-libraries"
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - E:\r_builds\25_2017-08-25_59654\ruby25_64\bin
     - C:\Windows\system32
     - C:\Windows
     - C:\Windows\System32\Wbem
     - C:\Windows\System32\WindowsPowerShell\v1.0\
     - C:\ruby25_64\bin
     - C:\Program Files\Git\cmd
     - E:\msys64
     - C:\Program Files (x86)\GNU\GnuPG\pub
     - C:\Program Files\Microsoft SQL Server\130\Tools\Binn\
     -
```

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by segiddins** [comment]
What is the expected behavior? What past version of RubyGems had that expected behavior?

**Comment by MSP-Greg** [comment]
[`gem update`](https://msp-greg.github.io/ruby_trunk/rubygems/file.command-reference.html#gem-update) & [`gem install`](https://msp-greg.github.io/ruby_trunk/rubygems/file.command-reference.html#gem-install) both describe `--explain` as follows:
```
-​-explain - Rather than install the gems, indicate which would be installed
```

Behavior:
```
E:\>gem list minitest

*** LOCAL GEMS ***

minitest (5.10.2)

E:\>gem update --explain
Updating installed gems
Nothing to update

E:\>gem update minitest --explain
Updating installed gems
Updating minitest
Fetching: minitest-5.10.3.gem (100%)
Successfully installed minitest-5.10.3
Gems updated: minitest
Gems already up-to-date: --explain
```

> What past version of RubyGems had that expected behavior?

Don't recall, otherwise I would have stated versions, etc.

**Comment by segiddins** [comment]
Yup, looks like `:explain` isn't handled at all in the update command, but adding it in seems more complicated than the implementation for install, so a PR would be appreciated!

**Comment by MSP-Greg** [comment]
I may have been mistaken about past behavior.  After reviewing command ref, I saw `outdated`, which may have been what I used in the past.  Re a PR: 

1. What should `gem install --explain` do?  I can see `gem install minitest --explain` as a valid cmd, but not one without gem names. Maybe a message like 'must specify a gem'.  Isn't an issue with `update`.

2. Overlap between `install`, `update`, and `outdated`.  For instance, `gem update --explain` (if left as is and fixed) is identical to `gem outdated`...

3. Stepping back, I think the three tasks below should be available:
  3.1 I don't have gem(s) X and Y  installed.  Which version(s) would be installed?
  3.2 I have gem(s) X and Y installed.  Which version(s) would be installed if I did `update`?
  3.3. Look thru **all my gems**, and show outdated ones (currently done by `outdated`.)

4. Given above, what & where, and whether `outdated` can/should allow a 'gem name list'

I'd be happy to look at a PR once exact functionality is pinned down.  I've used 'embedded ruby' in another app (no ruby.exe) and some rb scripts, so I'm somewhat familiar with `Gem::Commands` and `Gem::GemRunner`, but I wouldn't consider myself an expert.  Maybe swap for adding 2.3, 2.4. and trunk to Appveyor?

As an aside, I've got several versions of 2.3, 2.4, and trunk (I'm a builder).  Please take a look at [`did_you_mean`](https://rubygems.org/gems/did_you_mean).  Some build I have of 2.3 has 1.0.0 or 1.0.1 installed.  The most recent version(s) of `did_you_mean` have a ruby constraint for 2.4 or greater.  But, 1.0.3 does not.

When I did `gem update did_you_mean`, I got a message that it couldn't install 1.1.2.  Should it have installed 1.0.3?  I can recheck and file another issue if this is incorrect behavior.

**Comment by segiddins** [comment]
I've honestly never used `gem update`, so feedback from others on that would be helpful. Otherwise, maybe let's just remove the option from `update `since it's currently a no-op?

**Comment by MSP-Greg** [comment]
Ditto for `install --explain`?  Seems to have no affect; `:explain` just installs/updates the gem.
 
>I've honestly never used gem update

I assume for gem updates you use install?  

Any thoughts on whether RubyGems should 'backtrack' thru gem versions to find a version compatible with the version of Ruby one's using?  As in the `did_you_mean` example...

**Comment by segiddins** [comment]
`install --explain` seems to have it's own code path https://github.com/rubygems/rubygems/blob/master/lib/rubygems/commands/install_command.rb#L196

> I assume for gem updates you use install?

yup

> Any thoughts on whether RubyGems should 'backtrack' thru gem versions to find a version compatible with the version of Ruby one's using?

Ideally yes, but since Bundler already does that and RubyGems master can suggest a compatible version, I wouldn't recommend expending the effort on it

**Comment by MSP-Greg** [comment]
@segiddins

I'm okay to close this if you are...
